### ISSUE TITLE
REQUEST-33-APPLICATION-ATTACK-PHP.conf

### ISSUE THREAD

**Issue by emphazer** [SpiderLabs/owasp-modsecurity-crs#205]
Most PHP CMS Applications hide their .php 
the rules will not work if you match on the REQUEST.
Examples are Drupal or Typo3.
for example "https://www.drupal.org/support"
you have no match .php so the rule is useless.

i think it is necessary to let the rule work if its not a file...

i made a simple rule before it to fix that problem. 

SecRule REQUEST_URI "!^/.*\.[a-zA-Z]+" \
id:'400001',phase:1,t:none,nolog,noauditlog,pass,\
ctl:ruleRemoveById=958978

**Comment by csanders-git** [comment]
Hey emphazer,
Thanks for your feedback (keep it coming). In general URL rewriting is kind of an interesting assumption to generate rules around. While your rule solves the particular problem, this is a poor solution (in general) as it removes the entire reason for that particular rule (performance improvement). To state that another way, if you were using only ASP all PHP rules would still fire using your logic. I have thought over some methods for solving this such as setting a variable if we see 'powered by php' or requiring the user to set their languages as a variable during initial config. Tell me what you think. 

If you can come up with any ideas that seem to make sense drop by #modsecurity (irc.freenode.net) or just make a merge request. In general we're always happy with additional community support on  this project.

**Comment by emphazer** [comment]
Hi csanders, 
yeah, indeed thats a problem
but i think why should i load a PHP rule if i use ASP ??
working with ENV is a good idea but the result is the same.
i tried it with mod rewrite etc....

this works very well.
it scans everything without at dot and 1-6 signs at the end
so the perfomance waste should be quiet low.


SecRule REQUEST_FILENAME "!^/.*\.[^.]{1,6}$" \
id:'400001',phase:1,t:none,nolog,noauditlog,pass,\
ctl:ruleRemoveById=958978

**Comment by csanders-git** [comment]
You are absolutely correct. Users should all be like you ( :+1: ) and look to see what they are enabling and if it working. Unfortunately, we generally see users just enable everything and then complain :-1: . We are trying to cut down on that by automatically only including the rules they need. But you are right the rules as they stand don't make sense. In your situation, your rule makes perfect sense (and in fact we are considering removing rule 958978 based on your logic that users should only include it for php). We really do like when we get feedback so keep it coming and if you feel you want to make changes yourself you are more then free to request merges. Thanks!

**Comment by emphazer** [comment]
Sure thing ;-)
it would be a pleasure for me to help.

we use the CRS rules on round about >100 Rhel6 servers with jwall auditconsole and mlogc.
i made rpm packagages for CRS rules and mod_security.
the CRS rules are also patchable in my concept. 

the auditconsole server is a 128GB Ram Monster with 2x8 CPU *smile*

**Comment by csanders-git** [comment]
@emphazer I have been working on a 3.0.0 fork that solves some of these problems. The current working idea right now is that within the setup.conf we would have vars that would indicate the languages and protections that the user wants enabled (default all). Using this method we would actually not require people to manually choose what to include, they would set the options (although they could still remove .conf's for speed). Looks something like the following now

[Setup.conf]
  setvar:'tx.vulntypes=LFI RFI XSS SQLI LEAKAGE PROTOCOL FIXATION RCE',\
  setvar:'tx.languages=PHP JAVA ASP'"

[REQUEST-33]
SecRule TX:LANGUAGES "!@contains PHP" \
        "phase:request,\
        t:none,t:lowercase,\ ...
	skipAfter:END_PHP_APPLICATION_CHECKS"

Would a solution like this take care of your issues?

**Comment by emphazer** [comment]
Hi,

the following words are are listed twice:

##################################

lfi-os-files.txt

      2 /php/php.ini
      2 /var/log/mail.log


##################################

php-config-directives.txt

      2 error_log
      2 extension_dir
      2 include_path
      2 user_ini.filename
      3 session.save_path

##################################

php-errors.txt

      2 Encoding: SoapVar has no 'enc_type' property
      2 Invalid opcode 

##################################

php-function-names.txt

      2 eio_chmod
      2 eio_chown
      2 error_log
      2 ldap_connect
      2 mail
      2 mb_send_mail
      2 msession_connect
      2 mt_srand
      2 runkit_method_rename
      2 set_include_path
      2 sqlite_open
      2 sqlite_popen
      2 xdiff_file_bdiff
      2 xdiff_file_bpatch
      2 xdiff_file_diff
      2 xdiff_file_diff_binary
      2 xdiff_file_merge3
      2 xdiff_file_patch
      2 xdiff_file_patch_binary
      2 xdiff_file_rabdiff
      2 yaml_parse_file

##################################










i think i found a fault in id:958977

it seems that it HIT on "="
i could find any match in php-function-names.txt

Message: Warning. Matched phrase "=" at ARGS:body[und][0][value]. 
[file "/etc/httpd/modsecurity.d/activated_rules/REQUEST-33-APPLICATION-ATTACK-PHP.conf"]
[line "78"] 
[id "958977"] [rev "1"] 
[msg "PHP Injection Attack: Function Name Found"] 
[data "Matched Data: = found within ARGS:body[und][0][value]:
[severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "1"] [accuracy "8"] [tag "Host: test.ckan.nrw.de"] 
[tag "OWASP_CRS WEB_ATTACK/PHP_INJECTION"] [tag "OWASP_TOP_10/A1"]


###############################################################

your idea with the variables look nice. but i know if you load unnecessary rules it still drops perfomance.
let me give you an example. 
we had 2 big ssl proxy with a lot of accesses. its not possible to scan the outbout bacause the 4x 6cpus cant handle it.
i removed all outbound rules via SecRuleRemoveByTag and i set "SecResponseBodyAccess Off".
the server couldnt still handle the load... untill i really remove the outbound rules.

so i think the best solution would be to put special application rules in to an optional rule folder.
so anybody who really needs it could load them via symlink :-)

**Comment by csanders-git** [comment]
Hey @emphazer , this is great feedback. We can either fix those dups over here or you can do a pull request and then we'll merge it (this gives you the credit). Give me a little time to jump into that '=' issue i'll see whats causing it, if you haven't already. You are spot on with removing rules, let me think it over a bit :+1: ...

**Comment by emphazer** [comment]
Hi again!

i think i got it.
i saved the match of the first rule in a variable.


#
# [ PHP Configuration Directives ]
#
SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@pmf php-config-directives.txt" \
        "msg:'PHP Injection Attack: Configuration Directive Found',\
        phase:request,\
        rev:'1',\
        ver:'OWASP_CRS/3.0.0',\
        maturity:'1',\
        accuracy:'8',\
        capture,\
        t:none,t:normalisePath,\
        ctl:auditLogParts=+E,\
        block,\
        id:'958979',\
        tag:'OWASP_CRS/WEB_ATTACK/PHP_INJECTION',\
        tag:'OWASP_TOP_10/A1',\
        severity:'CRITICAL',\
        setvar:tx.php_injection_var=%{tx.0},\
        logdata:'Matched Data: %{TX.PHP_INJECTION_VAR}  found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
        chain"
                SecRule MATCHED_VARS "@pm =" \
                        "capture,\
                        setvar:'tx.msg=%{rule.msg}',\
                        setvar:tx.php_injection_score=+%{tx.critical_anomaly_score},\
                        setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},\
                        setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/PHP_INJECTION-%{matched_var_name}=%{tx.0}"




and then i received the output

Message: Warning. Matched phrase "=" at ARGS:test. [file "/etc/httpd/modsecurity.d/activated_rules/REQUEST-33-APPLICATION-ATTACK-PHP.conf"] [line "107"] [id "958979"] [rev "1"] [msg "PHP Injection Attack: Configuration Directive Found"] [data "Matched Data: session.use_trans_sid found within ARGS:test: session.use_trans_sid="] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "1"] [accuracy "8"]

**Comment by csanders-git** [comment]
Are you using a public app that triggers this? In general it is poor form to allow a user to control PHP configuration parameters via a parameter but if it is part of a commercial/OS app we can probably make an exception for this particular configuration value. I'm also going to push those changes  about the dups unless you wanted to do that :+1:

**Comment by emphazer** [comment]
im not very familiar with push requests and github.
would be great if you could do thath ;-)

**Comment by emphazer** [comment]
why are you asking about the public app?

**Comment by csanders-git** [comment]
If you have spare time I recommend giving https://gun.io/blog/how-to-github-fork-branch-and-pull-request/ a read, should answer any questions about merging into a git repo. In the mean time I merged your fixes.

In regards to it being a public app. If we saw something like wordpress undertaking this activity (session.use_trans_sid  being sent in a parameter) we might be inclined to white list it due to the frequency that that false positive might come up. This is what I meant by the question :). Also i'll soon be adding a check to the buildbots for uniqueness in data files based on your bug, good catch.
